fs/zonefs_zenfs.cc

```diff
@@ -0,0 +1,404 @@
+}
+
+uint64_t ZoneFsBackend::LBAToZoneOffset(uint64_t pos) {
+  return pos - ((pos / zone_sz_) * zone_sz_);
+}
+
+std::string ZoneFsBackend::GetBackingDevice(const char *mountpoint) {
+  struct mntent *mnt = NULL;
+  FILE *file = NULL;
+
+  file = setmntent("/proc/mounts", "r");
+  if (file == NULL) return "";
+
+  while ((mnt = getmntent(file)) != NULL) {
+    if (!strcmp(mnt->mnt_dir, mountpoint)) {
+      std::string dev_name(mnt->mnt_fsname);
+      std::size_t pos = dev_name.rfind("/");
+      if (pos != std::string::npos) {
+        dev_name.replace(0, pos + 1, "");
+        return dev_name;
+      }
+    }
+  }
+  endmntent(file);
+
+  return "";
+}
+
+unsigned int ZoneFsBackend::GetSysFsValue(std::string dev_name,
+                                          std::string field) {
+  std::ifstream sysfs;
+  unsigned int val = ZENFS_ZONEFS_DEFAULT_MAX_LIMIT;
+
+  sysfs.open("/sys/fs/zonefs/" + dev_name + "/" + field, std::ifstream::in);
+  if (sysfs.is_open()) {
+    sysfs >> val;
+    sysfs.close();
+  }
+
+  return val;
+}
+
+IOStatus ZoneFsBackend::Open(bool readonly,
+                             __attribute__((unused)) bool exclusive,
+                             unsigned int *max_active_zones,
+                             unsigned int *max_open_zones) {
+  struct stat zonefs_stat;
+
+  std::string seqdirname = mountpoint_ + "/seq";
+  if (stat(seqdirname.c_str(), &zonefs_stat) == -1) {
+    return IOStatus::InvalidArgument(
+        "Failed to access zonefs sequential zone directory: " +
+        ErrorToString(errno));
+  }
+
+  // The size of the "seq" directory shows the number of sequential zones
+  nr_zones_ = zonefs_stat.st_size;
+
+  seqdirname += "/0";
+  int zone_zero_fd = open(seqdirname.c_str(), O_RDONLY);
+  if (zone_zero_fd < 0) {
+    return IOStatus::InvalidArgument(
+        "Failed to open zonefs sequential zone 0: " + ErrorToString(errno));
+  }
+
+  if (flock(zone_zero_fd, LOCK_EX | LOCK_NB) == -1) {
+    close(zone_zero_fd);
+    return IOStatus::InvalidArgument(
+        "Failed to lock zonefs sequential zone 0: " + ErrorToString(errno));
+  }
+
+  if (fstat(zone_zero_fd, &zonefs_stat) == -1) {
+    close(zone_zero_fd);
+    return IOStatus::InvalidArgument(
+        "Failed to access zonefs sequential zone 0: " + ErrorToString(errno));
+  }
+  zone_sz_ = zonefs_stat.st_blocks * 512;
+  block_sz_ = zonefs_stat.st_blksize;
+
+  zone_zero_fd_ = zone_zero_fd;
+
+  std::string backing_dev =
+      ZoneFsBackend::GetBackingDevice(mountpoint_.c_str());
+  if (backing_dev.length()) {
+    *max_active_zones =
+        ZoneFsBackend::GetSysFsValue(backing_dev, "max_active_seq_files");
+    *max_open_zones =
+        ZoneFsBackend::GetSysFsValue(backing_dev, "max_wro_seq_files");
+  } else {
+    *max_active_zones = *max_open_zones = ZENFS_ZONEFS_DEFAULT_MAX_LIMIT;
+  }
+
+  wr_fds_.Resize(*max_active_zones);
+
+  readonly_ = readonly;
+
+  return IOStatus::OK();
+}
+
+std::unique_ptr<ZoneList> ZoneFsBackend::ListZones() {
```

Comment:
In an earlier revision of the patch set you were returning an IOStatus here, that allowed for better error reporting. How come you changed to pointer?